test/components/tools/test_tool_invoker.py

```diff
@@ -0,0 +1,195 @@
+
+        assert isinstance(tool_call_result, ToolCallResult)
+        assert tool_call_result.result == json.dumps({"weather": "mostly sunny", "temperature": 7, "unit": "celsius"})
+        assert tool_call_result.origin == tool_call
+
+    def test_run_with_invalid_message(self, invoker):
+        message_from_user = ChatMessage.from_user(text="Message from user.")
+        with pytest.raises(ValueError):
+            invoker.run(message=message_from_user)
+
+        message_wo_tool_calls = ChatMessage.from_assistant(text="Message without tool calls.")
+        with pytest.raises(ValueError):
+            invoker.run(message=message_wo_tool_calls)
+
+    def test_tool_not_found_error(self, invoker):
+        tool_call = ToolCall(
+            tool_name="non_existent_tool",
+            arguments={"location": "Berlin"}
+        )
+        tool_call_message = ChatMessage.from_assistant(
+            tool_calls=[tool_call]
+        )
+
+        with pytest.raises(ToolNotFoundException):
+            invoker.run(message=tool_call_message)
+
+    def test_tool_not_found_does_not_raise_exception(self, invoker):
+        invoker.raise_on_failure = False
+
+        tool_call = ToolCall(
+            tool_name="non_existent_tool",
+            arguments={"location": "Berlin"}
+        )
+        tool_call_message = ChatMessage.from_assistant(
+            tool_calls=[tool_call]
+        )
+
+        result = invoker.run(message=tool_call_message)
+        tool_message = result["tool_messages"][0]
+        assert "not found" in tool_message.tool_call_results[0].result
+
+    def test_tool_invocation_error(self, faulty_invoker):
+        tool_call = ToolCall(
+            tool_name="faulty_tool",
+            arguments={"location": "Berlin"}
+        )
+        tool_call_message = ChatMessage.from_assistant(
+            tool_calls=[tool_call]
+        )
+
+        with pytest.raises(ToolInvocationError):
+            faulty_invoker.run(message=tool_call_message)
+
+    def test_tool_invocation_error_does_not_raise_exception(self, faulty_invoker):
+        faulty_invoker.raise_on_failure = False
+
+        tool_call = ToolCall(
+            tool_name="faulty_tool",
+            arguments={"location": "Berlin"}
+        )
+        tool_call_message = ChatMessage.from_assistant(
+            tool_calls=[tool_call]
+        )
+
+        result = faulty_invoker.run(message=tool_call_message)
+        tool_message = result["tool_messages"][0]
+        assert "invocation failed" in tool_message.tool_call_results[0].result
+
+
+    def test_to_dict(self, invoker, weather_tool):
+        data = invoker.to_dict()
+        assert data == {
+            "type": "haystack_experimental.components.tools.tool_invoker.ToolInvoker",
+            "init_parameters": {
+                "tools": [weather_tool.to_dict()],
+                "raise_on_failure": True
+            },
+        }
+
+    def test_from_dict(self, weather_tool):
```

Comment:
Test that serializes/deserializes a pipeline with the component.

Reply:
done